src/libopensc/card-piv.c

```diff
@@ -2635,6 +2681,81 @@ static int piv_process_discovery(sc_card_t *card)
 	LOG_FUNC_RETURN(card->ctx, r);
 }
 
+/*
+ * parse a CCC to test  if this is a Dual CAC/PIV
+ * We read teh CCC using the PIV API.
+ * Look for CAC RID=A0 00 00 00 79
+ */
+ static int piv_parse_ccc(sc_card_t *card, u8* rbuf, size_t rbuflen)
+{
+	int r = 0;
+	const u8 * body;
+	size_t bodylen;
+
+	const u8 *p;
+	unsigned int cla_out, tag_out;
+
+	piv_private_data_t * priv = PIV_DATA(card);
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	if (rbuflen != 0)  {
+		body = rbuf;
+		if ((r = sc_asn1_read_tag(&body, rbuflen, &cla_out, &tag_out,  &bodylen)) != SC_SUCCESS) {
+			sc_log(card->ctx, "DER problem %d",r);
+			r = SC_ERROR_INVALID_ASN1_OBJECT;
+		goto err;
+		}
+	}
+
+	priv->ccc_flags |= PIV_CCC_FOUND;
+	p = body;
+	while (p < body + bodylen) {
+
+		switch (*p) {
+			case 0xf0:
+				if (*(p+1) == 0x15) {
+					if (memcmp((p+2) ,"\xA0\x00\x00\x03\08", 5) == 0)
+						priv->ccc_flags |= PIV_CCC_F0_PIV;
+					else if (memcmp((p+2) ,"\xA0\x00\x00\x00\x79", 5) == 0)
+						priv->ccc_flags |= PIV_CCC_F0_CAC;
+					if (*(p+8) == 0x02)
+						priv->ccc_flags |= PIV_CCC_F0_JAVA;
+				}
+				break;
+			case 0xf3:
+				if (*(p+1) == 0x10) {
+					if (memcmp((p+2) ,"\xA0\x00\x00\x00\x79\x04", 6) == 0)
+						priv->ccc_flags  |= PIV_CCC_F3_CAC_PKI;
+				}
+				break;
+		}
+		p = p + *(p+1) + 2;
+	}
+
+err:
+	LOG_FUNC_RETURN(card->ctx, r);
+}
```

Comment:
This function contains multiple possibilities for out of bounds reads!

Reply:
Fixed

Reply:
All this byte buffer dancing and length checking would be way more readable if you'd use `sc_asn1_read_tag` also for simple tags. See https://github.com/OpenSC/OpenSC/blob/a7a9326f35a2c46ba06fe480bf0d53b6f29edf91/src/libopensc/iso7816.c#L336-L419 for example.